src/predictors/schedule-based-predictor.cc

```diff
@@ -0,0 +1,191 @@
+      boost::geometry::set<0>(segment_start, stops[closest_segment_start].pos_.lng_);
+      boost::geometry::set<1>(segment_start, stops[closest_segment_start].pos_.lat_);
+      boost::geometry::set<0>(segment_end, stops[closest_segment_start + 1].pos_.lng_);
+      boost::geometry::set<1>(segment_end,
+                              stops[closest_segment_start + 1].pos_.lat_);
+      const double progress_way = boost::geometry::distance(segment_start, closest_foot_point, boost::geometry::strategy::distance::haversine(6371000.0)) /
+        boost::geometry::distance(segment_start, segment_end, boost::geometry::strategy::distance::haversine(6371000.0));
+      // progress_time: (current_time - a.departure_time) / (b.arrival_time - a.departure_time)
+      std::chrono::time_point<std::chrono::system_clock> now = std::chrono::system_clock::now();
+      const long current_time =
+          std::chrono::duration_cast<std::chrono::seconds>(
+              now.time_since_epoch())
+              .count();
+      const  nigiri::trip_idx_t trip_idx = predictorUtils::convert_trip_id_to_idx(timetable, tripID);
+      const nigiri::paged_vecvec<cista::strong<unsigned, nigiri::_trip_idx>, cista::pair<cista::strong<unsigned, nigiri::_transport_idx>, nigiri::interval<unsigned short>>>& transport_ranges = timetable.trip_transport_ranges_;
+      const cista::strong<unsigned, nigiri::_transport_idx> t_idx = transport_ranges[trip_idx][0].first; // Beachten Sie .from statt .from_
+      // convert to sys_days
+      const  std::chrono::time_point<std::chrono::system_clock, std::chrono::duration<int, std::ratio<86400>>> today = date::floor<date::days>(now);
+      const nigiri::day_idx_t current_day_idx = timetable.day_idx(today);
+      nigiri::unixtime_t departure_time = timetable.event_time(
+          nigiri::transport{
+              t_idx,
+              current_day_idx
+          },
+          // Convert from `location_idx_t` to `stop_idx_t`
+          nigiri::stop_idx_t{static_cast<unsigned short int>(
+              static_cast<unsigned>(closest_segment_start))},
+          nigiri::event_type::kDep
+      );
+      nigiri::unixtime_t arrival_time = timetable.event_time(
+          nigiri::transport{t_idx, current_day_idx},
+          // Convert from `location_idx_t` to `stop_idx_t`
+          nigiri::stop_idx_t{
+              static_cast<unsigned short int>(static_cast<unsigned>(
+                  closest_segment_start + 1))},
+          nigiri::event_type::kDep);
+      const double progress_time = (current_time - departure_time.time_since_epoch().count()) / (arrival_time.time_since_epoch().count() - departure_time.time_since_epoch().count());
+      // too_late = progress_time > progress_way
+      if (progress_time > progress_way) {
+        // Calculate predicted arrival: current_time + progress_time * (1 /
+        // progress_way)
+        const long predicted_arrival = current_time + static_cast<int>(progress_time * (1 / progress_way));
```

Comment:
I mean this should be `progress_time / progress_way` which is a relative value that we add to a time? I feel like this is not correct?

I would argue something like this makes more sense?
```suggestion
        const long predicted_arrival = current_time + static_cast<int>((scheduled_arrival_time - departure_time) * (1 - progress_way));
```
Here we would take the remaining progress and multiply it with the time we would still need in a normal situation. This would be an optimistic prediction, but I think makes more sense.

Reply:
Yes, you are right, thanks for the heads up! It is indeed a very basic prediction, but that's fine for now imho. I have also ideas on how to improve this (e.g. by using shapes of the routes instead of direct lines or by doing real routing, but this is too complex for the remaining time).